red4ext/src/invocable.rs

```diff
@@ -79,7 +79,15 @@ macro_rules! call {
         match $crate::call_direct!(
             rtti,
             $crate::types::RefShared::null(),
-            rtti.get_function($crate::types::CName::new($fn_name)),
+            match rtti.get_function($crate::types::CName::new($fn_name)) {
+                global if !global.is_null() => global,
+                _ => match ($fn_name.splitn(2, ':').next(), $fn_name.rsplitn(2, ':').next()) {
```

Comment:
re-splitting token that has just been concatenated here feels sad, but this is the only approach that has the least impact on the rest of the code, let me know if this is fine for you.

Reply:
does `rtti.get_function($crate::types::CName::new($fn_name))` return the actual function here?
if it does, it'd be a lot cleaner to just call `get_parent` on it:
```rs
$crate::rtti::Rtti::get_static_method(
  function.get_parent(),
  ffi::get_function_name(function)
)
```

you might need to add both get_parent and get_function_name:
```rs
// in #[namespace = "RED4ext"]
#[cxx_name = "GetParent"]
fn get_parent(self: Pin<&mut CBaseFunction>) -> *mut CClass;
```

```rs
// in #[namespace = "glue"]
#[cxx_name = "GetFunctionName"]
unsafe fn get_function_name(func: *const CBaseFunction) -> CName;
```

```cpp
// in namespace glue
CName GetFunctionName(const CBaseFunction* func) {
  return func->fullName;
}
```

Reply:
I'm not sure I get it because both  differs:

- original `rtti.get_function`:
   -  internally calls [RED4ext::IRTTISystem::GetFunction](https://github.com/WopsS/RED4ext.SDK/blob/master/include/RED4ext/RTTISystem.hpp#L26)
   > since it's a `virtual` and I'm not yet familiar with C++, I'm not sure where it's overriden with its concrete implementation (please let me know if you do).
   - returns a `CGlobalFunction*`
   - searches for e.g. `ClassName::FunctionName` (as per macro generated code)
   - is typically used with `redscript_global` on pure function
- added glue method `GetStaticMethod`
   - internally lookup in `cls->staticFuncs` and its parent
   - retrieves a `CClassStaticFunction*`
   - is required to search with separate parameters `ClassName` class and `FunctionName` function
      > ❌ searching for `ClassName::FunctionName` did not work when I tried
   - is typically used with `redscript_import` on `static func` defined on Reds `struct`
   
Does your suggestion allows for unifying both API ?

Reply:
What gets me confused is, when searching for `GetFunction` concrete implementation in `RED4ext.SDK` / `RED4ext` / `CET` / `Redscript` I can't seem to find it anywhere for `RTTISystem` or even `CRTTISystem`.

I can find implementations for e.g. `CClass`, but as far as I looked, it does not inherit from `RTTISystem`.

That's why, I'm not sure if it would lookup function name in e.g. `cls->funcs` as can be seen [here on Codeware](https://github.com/psiberx/cp2077-codeware/blob/main/lib/Red/TypeInfo/Invocation.hpp#L20), from both `cls->funcs` and `cls->staticFuncs` as can be seen for `CClass` [here on RED4ext.SDK](https://github.com/WopsS/RED4ext.SDK/blob/master/include/RED4ext/RTTITypes-inl.hpp#L370-L384), or any other way.

I would be grateful if you can shed a light on this,
my apologies for not being much familiar with C++ altogether.

Reply:
Ok in the meantime I gave a try to your suggestion and logged the results like so:

```rs
// in invocable.rs
// ...
([$fn_name:expr] ($( $args:expr ),*) -> $rett:ty) => {{
        let mut rtti = $crate::rtti::Rtti::get();
        let fully_qualified_name = $crate::types::CName::new($fn_name);
        let only_function_name = $crate::types::CName::new($fn_name.rsplitn(2, ':').next().unwrap());
        let mut fun = rtti.get_function(fully_qualified_name.clone());
        info!("rtti.get_function({:#?}|{:#?}) = {:#?}", fully_qualified_name, $crate::ffi::resolve_cname(&fully_qualified_name), fun);
        let mut fun = rtti.get_function(only_function_name.clone());
        info!("rtti.get_function({:#?}|{:#?}) = {:#?}", only_function_name, $crate::ffi::resolve_cname(&only_function_name), fun);
        if !fun.is_null() {
            let fun_name = unsafe { $crate::ffi::get_function_name(fun) };
            info!("ffi::get_function_name({:#?}|{:#?}) = {:#?}", fun, fun_name, $crate::ffi::resolve_cname(&fun_name));
            let mut parent = unsafe { fun.as_mut() };
            if parent.is_some() {
                let pinned = unsafe { core::pin::Pin::new_unchecked(parent.unwrap()) };
                let parent = pinned.get_parent();
                info!("({:#?}).get_parent({:#?})", fun, parent);
            }
        }
```
and here are the corresponding logs:
```txt
[2023-09-19 13:43:11.063] [example] [info] rtti.get_function(CName {
    hash: 6666660749396741679,
}|"") = 0x0000000000000000
[2023-09-19 13:43:11.068] [example] [info] rtti.get_function(CName {
    hash: 13569271457815481558,
}|"ToFloat") = 0x0000000000000000
```
so it seems the type (class) and the function must be resolved separately.

What do you think ?

Reply:
@Roms1383 I was confused about your change but I can see now that you're doing the lookup only if `get_function` returns null

I still don't like the idea of splitting on '::', there is a way of doing this without string manipulation:
- first bring back the `call_static` macro that I removed at some point because I thought it was redundant: https://github.com/jac3km4/red4ext-rs/commit/b97bf9aa813f1532e85bd1d4f11d11d874aaa443
- next, remove these lines in `generate_name` from `red4ext-macros/src/lib.rs` to remove the `...::` prefix:
  ```rs
    if let (None, Some(parent)) = (receiver, parent) {
        components.extend([quote!(#parent), quote!("::")]);
    }
  ```
- next, update code in `generate_forwarder` in the same file to handle forwarding to the static call macro:
  ```rs
    let body = match (receiver, parent) {
      (None, Some(class)) => {
          quote!(::red4ext_rs::call_static!(#class, #signature (#(#idents),*) #ret))
      }
      (Some(syn::Receiver { self_token, .. }), _) => {
  ```

That should be equivalent to your change

Reply:
@jac3km4 I gave a try and tested every methods, but alas it regresses with `[2023-09-20 10:07:31.116] [example] [error] Function 'CallDemo' has panicked: failed to invoke CanApplyBreathingEffect: function not found`.

It falls into the corner case where there's both no receiver and a defined class parent while e.g. :
- for an `IScriptable` like `PlayerPuppet`, it has to be called with `call!` + `PlayerPuppet::CanApplyBreathingEffect`
- but for a struct like `EngineTime`, it has to be called with `call_static!` + `ToFloat`

Essentially the macro doesn't know about the difference between `PlayerPuppet` and `EngineTime`, it sees a `static` so it just apply the same logic. And that's also why checking for the `nullptr` and falling back on getting class and function separately works currently.

Another option might be to pass `ClassName` and `FunctionName`, check if `FunctionName` exists, otherwise concatenate `ClassName::FunctionName` and check again if it exists. But it's still not very elegant and performant anyway.

> Now looking at @psiberx implementations in both [Red](https://github.com/psiberx/cp2077-red-lib) and [Codeware](https://github.com/psiberx/cp2077-codeware) libs, the proper way would probably be to refactor everything with a similar nicely split and orderly API, or even build bindings directly on top `Red` instead of `RED4ext.SDK`, but since this would be way more work I'd rather leave it for a future separate PR if you think this is the way to go.

Reply:
I think one way applies to native statics and the other way applies to scripted statics. You can check whether a function is native inside the procedural macro by checking the attributes so you can add a special case for native statics @Roms1383

Reply:
I'm still trying to find an elegant way because to check function flags it's still required to check whether `let mut fun = rtti.get_function(only_function_name.clone());` is `nullptr` or not in the first place. Maybe I can check class flags for `isScriptedClass` or `isScriptedStruct` tomorrow alongside, hopefully before Patch 2.0 kicks in.

Reply:
@Roms1383 the native attributes is already present in the procedural macro in `FunctionAttrs`. When a function is native the user is expected to provide the native flag to the macro, you can use it.

Reply:
Oh nice @jac3km4, indeed it's as easy as 8b4d674 😃